### DISCUSSION TITLE
pico to pico via local internet :  socket not accepting

### DISCUSSION THREAD

**Discussion by dondonddon** [micropython/micropython#13788] [RP2040 / Pico]
This should be easy (but apparently not)... I have 2 bare bones pico-w connected to my router.   The server socket won't accept the client 90% of the time.   Both server and client connect to the router well and the server ip address is always ('192.168.1.153', 80).  I can physically disconnect and reconnect both picos,  and most of the time, the client hangs because the server times out on the accept statement.   Any suggestions?

SERVER: 
```py
import socket
import secrets

import sys

led = machine.Pin('LED', machine.Pin.OUT)
#sequencing logic below:
#show that pico is active!
led.on()
sleep(1)

try:
    # ------------------  CONNECT -----------------
    print("Connect to router...")
    # Define SSID and password for the access point
    ssid = secrets.my_SSID
    password = secrets.my_word  
    wlan = network.WLAN(network.STA_IF)  
    wlan.active(True)
    wlan.connect(ssid, password)
    
     # Wait until it is active   
    while wlan.isconnected() == False:
        print('Waiting for connection...')
        sleep(1)
    print("Connected to router")
    ip = wlan.ifconfig()[0]
    print(f'Server connected on {ip}')
    print("WLAN,ip completed")
  
    # Open a socket
    address = (ip, 80)   #port 80
    print("Server address = ", address)
    connection = socket.socket()
    
    print("preparing connection for bind...")
    connection.bind(address)  
    print("Now listening for client on this connection:  ",connection)  
    connection.listen(1)  
    print("connection completed. Ready to serve")
    
    while True:
        try:
            client = connection.accept()[0] 
            print("Client connection accepted") 
            #Receive request from client
            request = client.recv(1024)
            print("Data received: ")
            request = str(request)
            print(f'Request received: {request}')
            #send back response:
            response = "OK"
            client.send(response.encode('UTF-8'))
            client.close()
            print("-"*30)
        except KeyboardInterrupt:
            print("Keyboard interrupt observed")
            #machine.reset()  #turns off access point. 
            sys.exit()
    
except OSError:
    print("Failure to connect.  Exiting..")
    print(OSError)

finally:
    connection.close()
    print("Finally: connection closed")
    sys.exit()
    #machine.reset()
```

CLIENT:
```py

from machine import Pin
from utime import sleep

import network
import socket
import secrets

import sys

led = machine.Pin('LED', machine.Pin.OUT)

#sequencing logic below:
#show that pico is active!
led.on()

try:
    # ------------------  CONNECT -----------------
    print("Connect to router...")
    ssid = secrets.my_SSID
    password = secrets.my_word  
    wlan = network.WLAN(network.STA_IF)
    wlan.active(True)
    wlan.connect(ssid, password)

    
     # Wait until it is active   
    while wlan.isconnected() == False:
        print('Waiting for connection...')
        sleep(1)
    print("Connected to router")
    ip = wlan.ifconfig()[0]
    print(f'Client connected on {ip}')
    print("WLAN,ip completed")
    
    while True:
        try:
            # Open a socket
            address = (ip, 80)   #port 80
            print("address = ", address)
            ai = socket.getaddrinfo('192.168.1.153', 80) # Address of Web Server
            print("ai : ", ai)
            addr = ai[0][-1]
            print("Server address ", addr)
            s = socket.socket()
            print("Client socket. Connecting to webserver...")
            s.connect(addr)  #bind socket to server ip address
            print("Now client and webserver connected")
            
            #request the time from the server:
            request = "request"
            s.send(request.encode('UTF-8'))
            print("Request: ", request)
            
            #Receive time:
            response = s.recv(1024) 
            print(f'Response received: {response}')
            response = response.decode("UTF-8")
            print("Response = ", response)
            s.close()
            print("-"*30)
            sleep(10) 
        except KeyboardInterrupt:
            print("Keyboard interrupt observed")
            #machine.reset()  #turns off access point.
            s.close()
            sys.exit()

except OSError:
    print("Failure to connect.  Exiting..")
    print(OSError)

finally:
    print("Client socket closing")
    s.close()
    sys.exit()
    #machine.reset()
    


```

**Comment by sophiedegran** [comment]
Hi 
Your code seem to be good...
Make a test with async examples:
github.com/peterhinch/micropython-async/blob/master/v3/as_driver/client_server/
You will find userver.py and uclient.py
Hope that help

**Reply by dondonddon** [reply]
Thanks.  I will look into the async version later this week.    I appreciate your help.

**Reply by dondonddon** [reply]
Sophie,  I did look at the github software today, trying to understand it.    I am somewhat comfortable with the asynch aspects of it.  I also like the general structure of the server and client.  But I am a bit perplexed as to not seeing code that I expected for both server and client connecting to a router (or the server being an access point for that matter).   So I am not certain how to implement it right now without some structural changes.   I will continue to look at it over the next day or two.    Also,  I did look at the some of the documentation associated with V3 and gained some good insight into the challenges (in the real world) of maintaining reliable long term connections.   It is well written and worth the time to peruse.

**Reply by dondonddon** [reply]
I did add in some code to have both Picos connected to my router and then use the ansyc code mentioned above.  Still have the connect problems between the server and router.   I am back to trying out the basic server/client codes, which are simpler and should be better able to find out why the server and client won't connect as  expected.  NOTE: sometimes the "simple server, simple client" do successfully connect and many times they don't.    Here is my latest codes to try to isolate and identify the problem...

SERVER:  
`
from machine import Pin
from time import sleep

import network
import socket
import secrets

import sys

led = machine.Pin('LED', machine.Pin.OUT)
#sequencing logic below:
#show that pico is active!
led.on()
sleep(1)
led.off()

try:
    # ------------------  CONNECT -----------------
    print("Connect to router...")
    # Define SSID and password for the access point
    ssid = secrets.my_SSID
    password = secrets.my_word  
    wlan = network.WLAN(network.STA_IF)  
    wlan.active(True)
    wlan.connect(ssid, password)
    
     # Wait until it is active   
    while wlan.isconnected() == False:
        print('Waiting for connection...')
        sleep(1)
    print("Connected to router")
    ip = wlan.ifconfig()[0]
    print(f'Server connected on {ip}')
    print("WLAN,ip completed")
    led.on()

except Exception as e:
    print("Failure to connect to router")
    print(e)
    led_off()
    sys.exit()
    

# Open a socket
address = (ip, 80)   #port 80
print("Server address = ", address)
connection = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
connection.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
connection.settimeout(90)
print("preparing connection for bind...")
connection.bind(address)  
print("Now listening for client on this connection:  ",connection)
connection.listen(1)  
print("connection completed. Ready to serve")

    
# ------------------ SERVE CLIENT  ----------------------   
while True:
    try:
        client = connection.accept()[0] 
        print("Client connection accepted")
    except Exception as e:
        print("Failure of client to connect to server")
        print(e)
        connection.close()
        led.off()
        sys.exit()
    #Receive request from client
    try:
        request = client.recv(1024)
        print("Request received: ", request)
        request = str(request.decode('UTF-8'))
        print(f'Decoded request received: {request}')
        #send back response:
        response = "OK"
        print("sending back response: ", response)
        client.send(response.encode('UTF-8'))
        client.close()
        print("-"*30)
    except KeyboardInterrupt:
        print("Keyboard interrupt")
        client.close()
        connection.close()
        led.off()
        sys.exit()



`

CLIENT:
`
from machine import Pin
from utime import sleep

import network
import socket
import secrets

import sys

led = machine.Pin('LED', machine.Pin.OUT)

#sequencing logic below:
#show that pico is active!
led.on()

try:
    # ------------------  CONNECT TO ROUTER -----------------
    print("Connect to router...")
    ssid = secrets.my_SSID
    password = secrets.my_word  
    wlan = network.WLAN(network.STA_IF)
    wlan.active(True)
    wlan.connect(ssid, password)   
    # Wait until it is active   
    while wlan.isconnected() == False:
        print('Waiting for connection...')
        sleep(1)
    print("Connected to router")
    ip = wlan.ifconfig()[0]
    print(f'Client connected on {ip}')
    print("WLAN,ip completed")
except Exception as e:
    print("Failure to connect to router")
    print(e)
    led.off()
    sys.exit()
    
sleep(5)
# ------------------  SOCKET FOR COMMUNICATION  -----------------
while True:
    # Open a socket
    print("address = ", ip)
    #ai = socket.getaddrinfo('192.168.1.153', 80) # Address of Web Server
    ai = socket.getaddrinfo('192.168.1.153', 80, 0, socket.SOCK_STREAM)  #  ref: https://docs.micropython.org/en/latest/library/socket.html
    print("ai : ", ai)
    addr = ai[0][-1]
    print("Server address ", addr)
        
    try:
        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        print("Client socket. Connecting to server...")
        s.settimeout(60)
        s.connect(addr)  #bind socket to server ip address
        print("Now client and server connected")
    except Exception as e:
        print("Failure of client to connect to server")
        print(e)
        led.off()
        s.close()
        sys.exit()
            
    #request to the server:
    try:
        request = "request"
        print("Sending this request: ", request)
        s.send(request.encode('UTF-8'))
        
        #Receive:
        response = s.recv(1024) 
        print(f'Response received: {response}')
        response_decoded = response.decode('UTF-8')
        print("Decoded response received = ", response_decoded)
        s.close()
        print("-"*30)
        sleep(10) 
    except KeyboardInterrupt:
        print("Keyboard interrupt observed")
        led.off()
        s.close()
        sys.exit()


`

**Comment by sophiedegran** [comment]
..
1- Normaly it's  good to have network connection under boot.py
And in main.py for one board the server and the other the client..
2- How do you know that your server have the good adress ip? did you see it on your router? 
(It seem that your server board ask simply a wifi connection(so dhcp)) , and for security reasons on your router you can't see (exchange) between 2 wifi connections..
Try first to isolate network..

**Reply by dondonddon** [reply]
I can see the IP address of the server on the router, so that is confirmed.   What I have NOT done is put the server code on the server pico board as main.py.  I have also NOT put the client code on the client pico as main.py.      I will try do do that in the next day or two (since it is late).   I believe I can see all the print statements in Thonny when I plug the boards into my USB socket.  That is key to observing the interaction between the server and client.   That is also WHY I have been running the software from Thonny with the boards simply attached.

**Reply by dondonddon** [reply]
FYI,  I tried the scenario where I have one pico-w board with main.py being the posted server code and the other pico-w board with its main.py being the client code.    I added 1 statement at the top : sleep(10) to give me 10 seconds to plug in the server board, then bring up Thonny so that I can see its output print statements.   When I see it waiting for a client, I plug in the other client board to another PC and bring up Thonny and watch its output.   As before,  I get the same result: both easily connect to the router, the server address remains 192.168.1.153 and the client tries to connect to it on port 80 and the connection attempts eventually  timeout.   So the bottom line appears to be that it doesn't matter if I run the code from Thonny  or have it on the board itself,  I still get (in genneral) the failure of that server/client connection.   I will start looking into some documentation on the router itself to see if there is something I need to configure in my ASUS 3100 router that might be not allowing these connection attempts.

**Reply by sophiedegran** [reply]
I think your router don't allow 'exchange' between wifi connections stations... You can install uping and test it..
Did your workstation see your pico server (assuming she's co. under wifi) ?
Go in advanced mode on your router, you'll find something I suppose
